tools/points-service/main.go

```diff
@@ -0,0 +1,778 @@
+        WHERE id = 1
+    `).Scan(&blk)
+	if err != nil {
+		return 0, err
+	}
+	return blk, nil
+}
+
+func (ps *PointsService) SetLastBlock(block uint64) error {
+	_, err := ps.db.Exec(`
+        UPDATE last_processed_block 
+        SET last_block = ? 
+        WHERE id = 1
+    `, block)
+	return err
+}
+
+// SetPointsRoutineRunning toggles the points routine status
+func (ps *PointsService) SetPointsRoutineRunning(running bool) {
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+	ps.pointsRoutineRunning = running
+}
+
+// IsPointsRoutineRunning returns the points routine status
+func (ps *PointsService) IsPointsRoutineRunning() bool {
+	ps.mu.RLock()
+	defer ps.mu.RUnlock()
+	return ps.pointsRoutineRunning
+}
+
+// SetSubscriptionActive toggles the event subscription status
+func (ps *PointsService) SetSubscriptionActive(active bool) {
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+	ps.eventSubscriptionActive = active
+}
+
+// IsSubscriptionActive returns the event subscription status
+func (ps *PointsService) IsSubscriptionActive() bool {
+	ps.mu.RLock()
+	defer ps.mu.RUnlock()
+	return ps.eventSubscriptionActive
+}
+
+func insertOptIn(db *sql.DB, logger *slog.Logger, pubkey, adder, registryType, eventType string, inBlock uint64) {
+	rwLock.RLock()
+	defer rwLock.RUnlock()
+
+	var existingAdder string
+	err := db.QueryRow(`
+        SELECT adder FROM validator_records
+        WHERE pubkey = ? AND opted_out_block IS NULL
+    `, pubkey).Scan(&existingAdder)
+
+	if err == nil && existingAdder != "" && existingAdder != adder {
+		logger.Warn("pubkey already opted in by a different adder",
+			"pubkey", pubkey, "existing_adder", existingAdder, "new_adder", adder)
+		return
+	}
+
+	_, err = db.Exec(`
+        INSERT INTO validator_records (
+            pubkey, adder, vault, registry_type, event_type, 
+            opted_in_block, opted_out_block, points_accumulated
+        ) VALUES (?, ?, NULL, ?, ?, ?, NULL, 0)
+    `, pubkey, adder, registryType, eventType, inBlock)
+	if err != nil {
+		logger.Warn("insertOptIn likely already inserted", "error", err)
+	} else {
+		logger.Info("inserted opt-in interval",
+			"pubkey", pubkey, "block", inBlock, "event_type", eventType, "adder", adder)
+	}
+}
+
+func insertOptInWithVault(db *sql.DB, logger *slog.Logger, pubkey, adder, vaultAddr, registryType, eventType string, inBlock uint64) {
+	rwLock.RLock()
+	defer rwLock.RUnlock()
+
+	var existingAdder string
+	err := db.QueryRow(`
+        SELECT adder FROM validator_records
+        WHERE pubkey = ? AND opted_out_block IS NULL
+    `, pubkey).Scan(&existingAdder)
+
+	if err == nil && existingAdder != "" && existingAdder != adder {
+		logger.Warn("pubkey already opted in by a different adder",
+			"pubkey", pubkey, "existing_adder", existingAdder, "new_adder", adder)
+		return
+	}
+
+	_, err = db.Exec(`
+        INSERT INTO validator_records (
+            pubkey, adder, vault, registry_type, event_type, 
+            opted_in_block, opted_out_block, points_accumulated
```

Comment:
This data model assumes a single validator will have one of each `opted_in_block` and `opted_out_block`. If they re-opt in a second time after opting in/out once, would a new row be created? Ie. how does the data model handle for this edge case? 

Note our contracts allow these sort of registration "cycles" for all actors. Ex: https://github.com/primev/mev-commit/blob/9c22b771b2792356ddd169bba07dca4ae8d581a1/contracts/test/validator-registry/middleware/MevCommitMiddlewareTest.sol#L1035

Reply:
Yes a new row is created.